adminSiteClient/SiteRedirectsIndexPage.tsx

```diff
@@ -0,0 +1,184 @@
+const INVALID_PATTERN_MESSAGE =
+    "URL cannot end with a slash unless it's the root."
+
+type Redirect = {
+    id: number
+    source: string
+    target: string
+}
+
+function RedirectRow({
+    redirect,
+    onDelete,
+}: {
+    redirect: Redirect
+    onDelete: (redirect: Redirect) => void
+}) {
+    return (
+        <tr>
+            <td className="text-break">{redirect.source}</td>
+            <td className="text-break">
+                <a href={redirect.target} target="_blank" rel="noopener">
+                    {redirect.target}
+                </a>
+            </td>
+            <td>
+                <button
+                    className="btn btn-danger"
+                    onClick={() => onDelete(redirect)}
+                >
+                    Delete
+                </button>
+            </td>
+        </tr>
+    )
+}
+
+type FormData = {
+    source: string
+    target: string
+}
+
+export default function SiteRedirectsIndexPage() {
+    const { admin } = useContext(AdminAppContext)
+    const [redirects, setRedirects] = useState<Redirect[]>([])
+    const [error, setError] = useState<string | null>(null)
+    const {
+        register,
+        formState: { errors, isSubmitting },
+        handleSubmit,
+        watch,
+    } = useForm<FormData>()
+    const source = watch("source")
+
+    useEffect(() => {
+        admin
+            .getJSON("/api/site-redirects.json")
+            .then((json) => {
+                setRedirects(json.redirects)
+            })
+            .catch(() => {
+                setError("Error loading redirects.")
+            })
+    }, [admin])
+
+    function validateTarget(value: string) {
+        if (value === source) {
+            return "Source and target cannot be the same."
+        }
+        return undefined
+    }
+
+    async function onSubmit(data: FormData) {
+        setError(null)
+        const json = await admin.requestJSON(
+            "/api/site-redirects/new",
+            data,
+            "POST"
+        )
+        if (json.success) {
+            setRedirects([json.redirect, ...redirects])
+        } else {
+            setError("Error creating a redirect.")
+        }
+    }
+
+    async function handleDelete(redirect: Redirect) {
+        if (!window.confirm(`Delete the redirect from ${redirect.source}?`))
+            return
+        setError(null)
+        const json = await admin.requestJSON(
+            `/api/site-redirects/${redirect.id}`,
+            {},
+            "DELETE"
+        )
+        if (json.success) {
+            setRedirects(redirects.filter((r) => r.id !== redirect.id))
+        } else {
+            setError("Error deleting the redirect.")
+        }
+    }
+
+    return (
+        <AdminLayout title="Site Redirects">
+            <main>
+                <form onSubmit={handleSubmit(onSubmit)}>
+                    <div className="form-row">
+                        <div className="form-group col-12 col-md-4">
```

Comment:
I think we should explain to the admin users a bit what is expected of them here. Yes, our admin UIs are very barebones but our authors are not very versed in the technical details of HTTP redirects so we should explain a bit what is happening here.

I think it is important to tell them that source has to start with a slash and should not contain query params or fragments (you may also want to strip those if they are given). The target can contain full urls or relative urls starting with a slash and here both query params and fragments are allowed. Maybe give an example for query param and fragment in parens as it might not be obvious what these things are.

Reply:
It is also worth calling out that to redirect to our homepage they should use a single slash as the target

Reply:
Added the instructions.